3 course/OperatingSystem/http-proxy/http_proxy.c

```diff
@@ -0,0 +1,177 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <arpa/inet.h>
+#include <netinet/in.h>
+#include <pthread.h>
+#include <netdb.h>
+
+#define MAX_BUFFER_SIZE 1024 * 4
+
+int setting_proxy_socket() {
+    struct sockaddr_in proxy_addr;
+    int proxy_socket;
+
+    proxy_socket = socket(AF_INET, SOCK_STREAM, 0);
+    if (proxy_socket < 0) {
+        perror("Ошибка при создании прокси-сокета");
+        exit(EXIT_FAILURE);
+    }
+
+    //настройка сокета для повторного использования порта
+    int opt = 1;
+    if (setsockopt(proxy_socket, SOL_SOCKET, SO_REUSEADDR, &opt, sizeof(opt))) {
+        perror("setsockopt");
+        exit(EXIT_FAILURE);
+    }
+
+    //настраиваем адрес и порт прокси-сокета
+    proxy_addr.sin_family = AF_INET;
+    proxy_addr.sin_addr.s_addr = INADDR_ANY;
+    proxy_addr.sin_port = htons(8080);
+
+    //привязываем прокси-сокет к адресу и порту
+    if (bind(proxy_socket, (struct sockaddr*)&proxy_addr, sizeof(proxy_addr)) < 0) {
+        perror("Ошибка при привязке прокси-сокета");
+        exit(EXIT_FAILURE);
+    }
+
+    //слушаем клиентиков
+    if (listen(proxy_socket, 5) < 0) {
+        perror("Ошибка при ожидании соединений");
+        exit(EXIT_FAILURE);
+    }
+
+    printf("[*] Прокси слушает на порту 80\n");
+
+    return proxy_socket;
+}
+
+void get_host(char* buffer, char* host) {
+    char buffer_copy[MAX_BUFFER_SIZE];
+    strcpy(buffer_copy, buffer);
+    char *first_line = strtok(buffer_copy, "\n");
+    char *url = strtok(first_line + 4, " ");
+    
+    int port;
+
+    sscanf(url, "http://%[^:/]:%d", host, &port);
+}
+
+int connect_and_get_server_socket(char *host) {
+    struct addrinfo hints, *result, *p;
+
+    memset(&hints, 0, sizeof(struct addrinfo));
+    hints.ai_family = AF_UNSPEC; 
+    hints.ai_socktype = SOCK_STREAM;
+
+    char* port = "80";
+    if (getaddrinfo(host, port, &hints, &result) != 0) {
+        perror("Ошибка при разрешении DNS");
+        exit(EXIT_FAILURE);
+    }
+
+    int server_socket;
+    for (p = result; p != NULL; p = p->ai_next) {
```

Comment:
according to code below you do not need a loop here. Please revise this place carefully